olive/passes/pytorch/pipeline_parallel.py

```diff
@@ -0,0 +1,183 @@
+# -------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+# --------------------------------------------------------------------------
+# Automatically distribute PyTorch model using Pipeline Parallelism
+# --------------------------------------------------------------------------
+
+import logging
+import multiprocessing
+from abc import abstractmethod
+from copy import deepcopy
+from pathlib import Path
+from typing import Any, Callable, Dict
+
+import torch
+
+from olive.common.config_utils import ParamCategory
+from olive.common.pydantic_v1 import validator
+from olive.hardware.accelerator import AcceleratorSpec, Device
+from olive.model import DistributedPyTorchModelHandler, PyTorchModelHandler
+from olive.model.config.hf_config import HfConfig, get_model_type_from_hf_config
+from olive.passes import Pass
+from olive.passes.olive_pass import PassConfigParam
+
+logger = logging.getLogger(__name__)
+
+
+class PipelineParallel:
+    def __init__(self, rank: int, world_size: int):
+        self.rank = rank
+        self.world_size = world_size
+
+    @abstractmethod
+    def replace_layers(self):
+        raise NotImplementedError
+
+    @abstractmethod
+    def restore_layers(self):
+        raise NotImplementedError
+
+    @abstractmethod
+    def split_layers(self, model: torch.nn.Module):
+        raise NotImplementedError
+
+
+class PyTorchPipelineParallel(Pass):
+    @staticmethod
+    def _default_config(accelerator_spec: AcceleratorSpec) -> Dict[str, PassConfigParam]:
+        # Note : The default value for nstages should be the no of gpus
+        # (AcceleratorSpec.Device == GPU) in the target OliveSystem
+        return {
+            "script_dir": PassConfigParam(
+                type_=str,
+                required=False,
+                category=ParamCategory.PATH,
+                description="Directory containing user script dependencies.",
+            ),
+            "user_script": PassConfigParam(
+                type_=str,
+                required=False,
+                category=ParamCategory.PATH,
+                description=(
+                    "Path to user script. The values for other parameters which were assigned"
+                    " function or object names will be imported from this script."
+                ),
+            ),
+            "class_name": PassConfigParam(
```

Comment:
the class_name seems not used?

![image](https://github.com/microsoft/Olive/assets/817030/c5a292df-abd9-4df3-b858-73ef5449ff10)

Reply:
Yes, there are few members in both tensor parallel & pipeline parallel implementation that need to be removed. I will do a follow up PR to clean it up.